File: examples/example_client.js
#!/usr/bin/env node
/**
 * Example Node.js/JavaScript client for Live Audio Stream Transcription
 * Demonstrates how to use the WebSocket API programmatically
 * 
 * Installation:
 *   npm install ws
 * 
 * Usage:
 *   node example_client.js <url> [language]
 *   node example_client.js https://example.com/audio.mp3
 *   node example_client.js https://example.com/video.m3u8 en
 */

const WebSocket = require('ws');

// Configuration
const WS_URL = 'ws://localhost:8000/ws/transcribe';

/**
 * Transcribe an audio/video stream URL
 */
async function transcribeStream(url, language = null) {
    return new Promise((resolve, reject) => {
        console.log(`\nConnecting to ${WS_URL}...`);
        
        const ws = new WebSocket(WS_URL);
        const fullTranscription = [];
        
        // Connection opened
        ws.on('open', () => {
            console.log('✓ Connected to transcription service\n');
            
            // Send transcription request
            const request = {
                url: url,
                language: language
            };
            
            console.log(`Starting transcription for: ${url}`);
            if (language) {
                console.log(`Language: ${language}`);
            }
            console.log('='.repeat(60));
            
            ws.send(JSON.stringify(request));
        });
        
        // Receive messages
        ws.on('message', (data) => {
            const message = JSON.parse(data.toString());
            
            if (message.type === 'status') {
                console.log(`Status: ${message.message}`);
            }
            else if (message.type === 'transcription') {
                const text = message.text;
                const detectedLang = message.language || 'unknown';
                
                console.log(`\n[${detectedLang}] ${text}`);
                fullTranscription.push(text);
            }
            else if (message.type === 'complete') {
                console.log('\n' + '='.repeat(60));
                console.log('✓ Transcription complete!');
                
                if (fullTranscription.length > 0) {
                    console.log('\n' + '='.repeat(60));
                    console.log('FULL TRANSCRIPTION:');
                    console.log('='.repeat(60));
                    console.log(fullTranscription.join(' '));
                    console.log('='.repeat(60));
                    console.log(`\nTotal segments: ${fullTranscription.length}`);
                    console.log(`Total words: ${fullTranscription.join(' ').split(/\s+/).length}`);
                }
                
                ws.close();
                resolve(fullTranscription);
            }
            else if (message.error) {
                console.error(`\n❌ Error: ${message.error}`);
                ws.close();
                reject(new Error(message.error));
            }
        });
        
        // Handle errors
        ws.on('error', (error) => {
            console.error(`❌ WebSocket error: ${error.message}`);
            if (error.code === 'ECONNREFUSED') {
                console.error('Connection refused. Is the server running on port 8000?');
                console.error('Start the server with: docker-compose up');
            }
            reject(error);
        });
        
        // Handle close
        ws.on('close', () => {
            console.log('\nConnection closed');
        });
    });
}

/**
 * Main function
 */
async function main() {
    console.log('='.repeat(60));
    console.log('Live Audio Stream Transcription - JavaScript Client');
    console.log('='.repeat(60));
    
    // Parse command line arguments
    const args = process.argv.slice(2);
    
    if (args.length === 0) {
        console.log('\nUsage:');
        console.log('  node example_client.js <url> [language]');
        console.log('\nExamples:');
        console.log('  node example_client.js https://example.com/audio.mp3');
        console.log('  node example_client.js https://example.com/stream.m3u8 en');
        console.log('  node example_client.js https://example.com/video.mp4 es');
        console.log('\nSupported languages:');
        console.log('  en (English), es (Spanish), fr (French), de (German),');
        console.log('  it (Italian), pt (Portuguese), ru (Russian), ja (Japanese),');
        console.log('  ko (Korean), zh (Chinese), ar (Arabic), hi (Hindi), etc.');
        console.log('\nNote: Language is optional. Auto-detection will be used if not specified.\n');
        process.exit(1);
    }
    
    const url = args[0];
    const language = args[1] || null;
    
    try {
        await transcribeStream(url, language);
        console.log('\n✓ Done!\n');
        process.exit(0);
    } catch (error) {
        console.error(`\n❌ Failed: ${error.message}\n`);
        process.exit(1);
    }
}

// Run if executed directly
if (require.main === module) {
    main().catch((error) => {
        console.error('Fatal error:', error);
        process.exit(1);
    });
}

// Export for use as module
module.exports = { transcribeStream };

File: examples/example_client.py
#!/usr/bin/env python3
"""
Example Python client for Live Audio Stream Transcription
Demonstrates how to use the WebSocket API programmatically
"""
import asyncio
import websockets
import json
import sys

async def transcribe_stream(url: str, language: str = None):
    """
    Transcribe an audio/video stream URL
    
    Args:
        url: Audio or video stream URL (m3u8, mp4, mp3, etc.)
        language: Optional language code (e.g., 'en', 'es', 'fr')
    """
    # WebSocket endpoint
    ws_url = "ws://localhost:8000/ws/transcribe"
    
    print(f"Connecting to {ws_url}...")
    
    try:
        async with websockets.connect(ws_url) as websocket:
            print("✓ Connected to transcription service")
            
            # Send transcription request
            request = {
                "url": url,
                "language": language
            }
            
            print(f"\nStarting transcription for: {url}")
            if language:
                print(f"Language: {language}")
            print("\n" + "="*60)
            
            await websocket.send(json.dumps(request))
            
            # Receive and process transcription results
            full_transcription = []
            
            async for message in websocket:
                data = json.loads(message)
                
                if data.get("type") == "status":
                    print(f"Status: {data['message']}")
                    
                elif data.get("type") == "transcription":
                    text = data['text']
                    detected_lang = data.get('language', 'unknown')
                    
                    print(f"\n[{detected_lang}] {text}")
                    full_transcription.append(text)
                    
                elif data.get("type") == "complete":
                    print("\n" + "="*60)
                    print("✓ Transcription complete!")
                    break
                    
                elif data.get("error"):
                    print(f"\n❌ Error: {data['error']}")
                    break
            
            # Print full transcription
            if full_transcription:
                print("\n" + "="*60)
                print("FULL TRANSCRIPTION:")
                print("="*60)
                print(" ".join(full_transcription))
                print("="*60)
                print(f"\nTotal segments: {len(full_transcription)}")
                print(f"Total words: {len(' '.join(full_transcription).split())}")
            
    except websockets.exceptions.ConnectionRefused:
        print("❌ Connection refused. Is the server running on port 8000?")
        print("   Start the server with: docker-compose up")
        sys.exit(1)
        
    except Exception as e:
        print(f"❌ Error: {e}")
        sys.exit(1)


async def main():
    """Main function"""
    print("="*60)
    print("Live Audio Stream Transcription - Python Client")
    print("="*60)
    print()
    
    # Example URLs - replace with your own
    example_urls = {
        "1": {
            "url": "https://www.soundhelix.com/examples/mp3/SoundHelix-Song-1.mp3",
            "name": "Direct MP3 Audio",
            "language": None
        },
        "2": {
            "url": "https://test-streams.mux.dev/x36xhzz/x36xhzz.m3u8",
            "name": "HLS Stream (m3u8)",
            "language": None
        },
    }
    
    # Interactive selection or command-line argument
    if len(sys.argv) > 1:
        # URL provided as command-line argument
        url = sys.argv[1]
        language = sys.argv[2] if len(sys.argv) > 2 else None
        await transcribe_stream(url, language)
    else:
        # Interactive mode
        print("Select an option:")
        print("1. Test with example MP3 audio")
        print("2. Test with example m3u8 stream")
        print("3. Enter custom URL")
        print()
        
        choice = input("Enter choice (1-3): ").strip()
        
        if choice in example_urls:
            example = example_urls[choice]
            print(f"\nUsing: {example['name']}")
            await transcribe_stream(example['url'], example['language'])
        elif choice == "3":
            url = input("\nEnter audio/video URL: ").strip()
            language = input("Enter language code (optional, press Enter to skip): ").strip() or None
            await transcribe_stream(url, language)
        else:
            print("Invalid choice")
            sys.exit(1)


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n\nInterrupted by user")
        sys.exit(0)
